fs/fcb/src/fcb_getnext.c

```diff
@@ -51,63 +90,377 @@ fcb_getnext_area(struct fcb *fcb, struct flash_area *fap)
+        cache->sector_ix = 0xFFFF;
+    }
+}
+
+/**
+ * Function invalidates cache if sector changes
+ *
+ * If sector number is different that one that was cached all entries are
+ * removed.
+ *
+ * @param fcb - fcb to invalidate cache
+ * @param cache - cache to invalidate if needed
+ * @param sector_ix - sector number
+ */
+static void
+fcb_cache_switch_to_sector(struct fcb *fcb, struct fcb_entry_cache *cache, int sector_ix)
+{
+    if (cache != NULL && cache->sector_ix != sector_ix) {
+        fcb_cache_invalidate(fcb, cache);
+        cache->sector_ix = sector_ix;
+    }
+}
+
+/**
+ * Retrieve element length from cache
+ *
+ * @param fcb - fcb to use
+ * @param cache - cache to use
+ * @param elem_ix - element index
+ * @return element length if present in cache
+ *         SYS_NOENT if cache does not have element
+ */
+static int
+fcb_cache_elem_len(struct fcb *fcb, struct fcb_entry_cache *cache, int elem_ix)
+{
+    int len;
+
+    (void)fcb;
+
+    if (cache && cache->cache_count > elem_ix) {
+        len = cache->cache_data[elem_ix];
     } else {
-        rc = fcb_getnext_in_area(fcb, loc);
+        len = SYS_ENOENT;
     }
-    switch (rc) {
-    case 0:
-        return 0;
-    case FCB_ERR_CRC:
-        break;
-    default:
-        goto next_sector;
+    return len;
+}
+
+/**
+ * Store entry information in cache.
+ *
+ * Stores entry
+ * @param fcb - fcb to use
+ * @param cache - cache to store information (can be NULL)
+ * @param entry_ix - index of entry in cache
+ * @param offset - offset of entry
+ * @param elem_size - FCB entry size in bytes
+ */
+static void
+fcb_cache_add(struct fcb *fcb, struct fcb_entry_cache *cache, int entry_ix, int offset, int elem_size)
+{
+    /* Total bytes for entries in sector */
+    int sector_data_size;
+    int first_entry_offset;
+
+    if (cache == NULL || cache->cache_data == NULL) {
+        return;
     }
-    while (rc == FCB_ERR_CRC) {
-        rc = fcb_getnext_in_area(fcb, loc);
-        if (rc == 0) {
-            return 0;
+
+    first_entry_offset = fcb_start_offset(fcb);
+
+    /* Data size is size of sector minus sector header */
+    sector_data_size = fcb_sector_size(fcb, cache->sector_ix) - first_entry_offset;
+
+    if (entry_ix >= cache->cache_size) {
+        assert(cache->cache_count == cache->cache_size);
+
+        int average_entry_size = (offset - first_entry_offset) / entry_ix;
+        /*  guess cache size based on entries currently in flash */
+        int new_size = cache->cache_size + 5 + (sector_data_size - offset) / average_entry_size;
+        size_t cache_byte_count = new_size * sizeof(cache->cache_data[0]);
+
+        cache->cache_data = os_realloc(cache->cache_data, cache_byte_count);
+        if (cache->cache_data == NULL) {
+            cache->sector_ix = 0xFFFF;
+            cache->cache_size = 0;
+            cache->cache_count = 0;
+        } else {
+            cache->cache_size = new_size;
         }
+    }
+
+    if (cache->cache_data) {
```

Comment:
this could use some comment:
elem_size is converted to host, but other params (ix, offset) are used as is
also elem_size is (32bit size?) int but it is stored into cache_data which is uint16

If cache is kept in host order, maybe this function should get all params in host order?

Reply:
inconsistent endianness removed.
function argument change to uint16_t to match calling argument